### ISSUE TITLE
Training

### ISSUE THREAD

**Issue by abhiML** [bharatv007/Lung-Cancer-Detection-Kaggle#1]
Hey @bharatv007 
what is loss and dice coefficient you get after training the Luna16 dataset with your script? 
Did you make any changes to the DSB3Tutorial code in your repo?

**Comment by bharatv007** [comment]
Hi,
I think I have uploaded an older version of my code. The major changes I
made were in normalising the images. If I recall, the algorithm takes in a
4 dimensional array for detecting the lung nodules.
1)Make sure you normalise it along the right axis.
2) You have to deal with overfitting.
3) you can read the report that is attached in the GitHub repo

Cheers,
Bharat V

**Comment by abhiML** [comment]
Thanks for the reply.What exactly do you mean by normalising along the right axis? The imgs_train(input to the UNET) is a 4D array which is 
nx1x512x512, where n is the number of training image files.

**Comment by bharatv007** [comment]
When you normalise, you need to take the mean of each image(512x512) and
subtract for each image. So technically, you should get nx1.

**Comment by abhiML** [comment]
Thanks but are we not doing that when we are subtracting the mean from the images and then dividing by the "std" before passing the images to the UNET?

**Comment by abhiML** [comment]
Also regarding overfitting I saw that you had added dropout layers in the code that you had uploaded. Are there any other steps that you had done which is not there in the code?

**Comment by bharatv007** [comment]
1)Try to save only the best weights after each epoch.
2)Use drop outs in the deep learning architecture.
3) Create a validation set and measure the error on the same. Make sure u
have low variance and low bias when you train. When you reach a decent
error(90%) on the validation set. Stop the training and start testing.

**Comment by abhiML** [comment]
And regarding the normalisation of the images what exactly do I do? The test as well as the train images have been normalised using the mean and std. Is there anything else to be done?

**Comment by bharatv007** [comment]
This normalisation does give you good results. You can also try other
methods, like just divide each pixel by 255 or use a min-max normalisation.
All these perform good, theoretically. But Deep learning is very prone to
such normalisation and different techniques may lead to faster convergence.
If you have the resources and time, ideally thing to do would be to try
many normalisation techniques.

**Comment by abhiML** [comment]
Right. I saw in your report you ran the training for around 70 epochs. Also did you use the entire dataset?

**Comment by abhiML** [comment]
I am using subset0-6. The Dice coefficient is plateauing at 0.02 to 0.03 from epoch 6 to epoch 172(running right now). How many epochs does it take to converge? Is this behaviour normal?

**Comment by abhiML** [comment]
And did you normalise the masks or just the lung images?

**Comment by bharatv007** [comment]
I do not recall exactly what I did. But I guess reshaping the images is not
a good idea. It just scales it in a wrong way( you can choose to omit this
step and try it). The dice coefficient has to be higher I think. For basic
performance you can try it with a small data set and run it for 50-75
epochs. Check the performance and then see if there is a bug in code. Then
you can use the whole data set, so that the model generalises well. The
code that is uploaded has lots of bugs in it. My colleague is working on
the same project for his research(extension), so I can not upload the
version that is more or less complete( some methods that I tried).  Once he
allows me, I can upload my code!

**Comment by abhiML** [comment]
Could you at least upload a trained model for the task? I require it for a presentation rather urgently.

**Comment by bharatv007** [comment]
I do not want to do that. This is a hard project. The fact that you
attempted this project is in itself a huge deal. So keep trying, I am sure
you can crack it. Spend some time with the code. I can always help
you(guide you), but I will not upload the weights.

**Comment by abhiML** [comment]
Alright :) I'll keep at it and get back to you.

**Comment by abhiML** [comment]
Okay :)
So what exactly is causing the bugs in the code? If I simply gave the UNET the segmented lungs and the corresponding nodule masks why will it not be able to converge? It should considering the UNET is used for many segmentation tasks.

**Comment by bharatv007** [comment]
Yes. Make sure the pipeline is right. You need to make sure you pre-process
steps are same for training and test set. If you have any specific error or
difficulties, it will be easier for me to suggest you tricks to overcome
them.
